### DISCUSSION TITLE
Why do static functions share expectations?

### DISCUSSION THREAD

**Discussion by 128f** [asomers/mockall#361] [Questions]
[Over here](https://github.com/cr0ax/mock-tests) I did a test and saw that the tests fail because of the shared expectations.

Is this a design choice, or a technical limitation? If it's the latter, can we find a way to solve it?

**Comment by asomers** [comment]
static functions' expectations are global, because there's no mock object to own them.  With non-static functions, the expectation state is owned by the mock object.  But static functions don't have any associated mock object.  So you need to synchronize your tests with some kind of Mutex.  Here's an example of how to do that: https://github.com/asomers/mockall/blob/master/mockall/examples/synchronization.rs .

**Reply by 128f** [reply]
Thanks. Are there plans to create this kind of functionality inside the generated code?

**Reply by asomers** [reply]
It's been discussed.  However, adding that Mutex internally could lead to an even worse problem: deadlocks.  Deadlocks would be very surprising to users who didn't even know that the Mutex existed.
